File: ontology-engine/graph-engine/src/main/java/org/sunbird/graph/engine/DataNode.java
package org.sunbird.graph.engine;

import akka.dispatch.Futures;
import akka.dispatch.Mapper;
import akka.pattern.Patterns;
import org.apache.commons.lang3.StringUtils;
import org.sunbird.common.dto.Request;
import org.sunbird.common.dto.Response;
import org.sunbird.common.exception.ResponseCode;
import org.sunbird.graph.dac.model.Node;
import org.sunbird.graph.dac.model.Relation;
import org.sunbird.graph.engine.dto.Result;
import org.sunbird.graph.mgr.BaseGraphManager;
import org.sunbird.graph.service.operation.Neo4JBoltNodeOperations;
import scala.concurrent.Future;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;


public class DataNode extends BaseDomainObject {

    private BaseGraphManager manager;

    public DataNode(BaseGraphManager manager, String graphId, String objectType, String version) {
        super(graphId, objectType, version);
        this.manager = manager;
    }

    public void create(Request request) throws Exception {
        Map<String, Object> data = request.getRequest();
        Result validationResult = validate(objectType, data);

        if(validationResult.isValid()) {
            Response response = createNode(validationResult.getNode());
            Future<Object> extPropsResponse = saveExternalProperties(validationResult.getExternalData(), request.getContext());
            Future<Object> updateRelResponse = updateRelations(Arrays.asList(), request.getContext());
            List<Future<Object>> futureList = Arrays.asList(extPropsResponse, updateRelResponse);
            // TODO: onComplete of futureList, if all successful - send the response.
            Future<Iterable<Object>> future = Futures.sequence(futureList, manager.getContext().getDispatcher());
            Future<Object> result = future.map(new Mapper<Iterable<Object>, Object>() {
                @Override
                public Object apply(Iterable<Object> f) {
                    List<Response> resList = StreamSupport.stream(f.spliterator(), false)
                            .filter(res -> {
                        String code = ((Response) res).getResponseCode().name();
                        return !StringUtils.equals(code, ResponseCode.OK.name());
                    }).map(res -> (Response) res).collect(Collectors.toList());
                    if (resList.size() == 0) {
                        return response;
                    } else {
                        return resList.get(0);
                    }
                }
            }, manager.getContext().dispatcher());
            Patterns.pipe(result, manager.getContext().dispatcher()).to(manager.sender());
        } else {
            Response response = new Response();
            response.setResponseCode(ResponseCode.CLIENT_ERROR);
            response.put("messages", validationResult.getMessages());
            Patterns.pipe(Futures.successful(response), manager.getContext().dispatcher()).to(manager.sender());
        }


    }

    private Result validate(String objectType, Map<String, Object> data) throws Exception {
        DefinitionNode definition = new DefinitionNode(graphId, objectType, "1.0");
        return definition.validate(data);
    }

    private Response createNode(Node node) {
        Node addedNode = Neo4JBoltNodeOperations.addNode(graphId, node);
        Response response = new Response();
        response.put("node_id", addedNode.getIdentifier());
        response.put("versionKey", addedNode.getMetadata().get("versionKey"));
        return response;
    }

    private Future<Object> saveExternalProperties(Map<String, Object> externalProps, Map<String, Object> context) {
        // TODO: Update externalProps using ExternalStoreActor (Generic CassandraStorage).
        Request request = new Request(context, externalProps, "saveExternalProperties", objectType);
        return manager.getResult(request);
    }

    private Future updateRelations(List<Relation> relations, Map<String, Object> context) {
        // TODO: update Relations using GraphManager Actor.
        return Futures.successful(new Response());
    }
}

File: ontology-engine/graph-engine_2.11/src/main/scala/org/sunbird/graph/nodes/DataNode.scala
package org.sunbird.graph.nodes

import java.util

import org.apache.commons.collections4.{CollectionUtils, MapUtils}
import org.sunbird.common.JsonUtils
import org.sunbird.common.dto.{Request, Response}
import org.sunbird.graph.dac.model.{Node, Relation}
import org.sunbird.graph.engine.RelationManager
import org.sunbird.graph.external.ExternalPropsManager
import org.sunbird.graph.relations.{IRelation, RelationHandler}
import org.sunbird.graph.schema.{DefinitionFactory, DefinitionNode}
import org.sunbird.graph.service.operation.NodeAsyncOperations
import org.sunbird.parseq.Task

import scala.collection.JavaConversions._
import scala.concurrent.{ExecutionContext, Future}


object DataNode {
    @throws[Exception]
    def create(request: Request)(implicit ec: ExecutionContext): Future[Node] = {
        val graphId: String = request.getContext.get("graph_id").asInstanceOf[String]
        val version: String = request.getContext.get("version").asInstanceOf[String]
        val definition = DefinitionFactory.getDefinition(graphId, request.getObjectType, version)
        val inputNode = definition.getNode(request.getRequest)
        val validatedNode = definition.validate(inputNode, "create")
        validatedNode.map(node => {
            val response = NodeAsyncOperations.addNode(graphId, node)
            response.map(result => {
                val futureList = Task.parallel[Response](
                    saveExternalProperties(node.getIdentifier, node.getExternalData, request.getContext, request.getObjectType),
                    updateRelations(graphId, node, request.getContext))
                futureList.map(list => result)
            }).flatMap(f => f)
        }).flatMap(f => f)
    }

    @throws[Exception]
    def update(request: Request)(implicit ec: ExecutionContext): Future[Node] = {
        val graphId: String = request.getContext.get("graph_id").asInstanceOf[String]
        val version: String = request.getContext.get("version").asInstanceOf[String]
        val identifier: String = request.getContext.get("identifier").asInstanceOf[String]
        val definition = DefinitionFactory.getDefinition(graphId, request.getObjectType, version)
        val dbNodeFuture = definition.getNode(identifier, "update", null)
        val validationResult: Future[Node] = dbNodeFuture.map(dbNode => {
            val inputNode: Node = definition.getNode(dbNode.getIdentifier, request.getRequest, dbNode.getNodeType)
            val validatedNode = definition.validate(inputNode, "update")
            validatedNode.map(dataNode => {
                dbNode.getMetadata.putAll(dataNode.getMetadata)
                dbNode.getInRelations.addAll(dataNode.getInRelations)
                dbNode.getOutRelations.addAll(dataNode.getOutRelations)
                dbNode.getAddedRelations.addAll(dataNode.getAddedRelations)
                dbNode.getDeletedRelations.addAll(dataNode.getDeletedRelations)
                dbNode.setRelationNodes(dataNode.getRelationNodes)
                dbNode.setExternalData(dataNode.getExternalData)
                dbNode
            })
        }).flatMap(f => f)

        validationResult.map(dbNode => {
            val response = NodeAsyncOperations.upsertNode(graphId, dbNode, request)
            response.map(result => {
                val futureList = Task.parallel[Response](
                    saveExternalProperties(dbNode.getIdentifier, dbNode.getExternalData, request.getContext, request.getObjectType),
                    updateRelations(graphId, dbNode, request.getContext))
                futureList.map(list => result)
            }).flatMap(f => f)
        }).flatMap(f => f)
    }

    private def saveExternalProperties(identifier: String, externalProps: util.Map[String, AnyRef], context: util.Map[String, AnyRef], objectType: String)(implicit ec: ExecutionContext): Future[Response] = {
        if (MapUtils.isNotEmpty(externalProps)) {
            externalProps.put("identifier", identifier)
            val request = new Request(context, externalProps, "", objectType)
            ExternalPropsManager.saveProps(request)
        } else {
            Future(new Response)
        }
    }
    
    private def updateRelations(graphId: String, node: Node, context: util.Map[String, AnyRef])(implicit ec: ExecutionContext) : Future[Response] = {
        val relations: util.List[Relation] = node.getAddedRelations
        if (CollectionUtils.isNotEmpty(relations)) {
            val relationList: List[IRelation] = relations.toList.map(relation =>
                RelationHandler.getRelation(graphId, node.getRelationNode(relation.getStartNodeId),
                    relation.getRelationType, node.getRelationNode(relation.getEndNodeId), new util.HashMap()))
            val request: Request = new Request
            request.setContext(context)
            request.put("relations", relationList)
            RelationManager.createNewRelations(request)
        } else {
            Future(new Response)
        }
    }
}